### ISSUE TITLE
Fractional days cause rotations start_dates and plots to accumulate errors

### ISSUE THREAD

**Issue by woodbri** [farmOS/farm_grazing#4]
Fractional grazing days get rounded, so when there is a long list of paddocks in the rotations table the plotted start times are off the further down the table as you accumulate more errors.

For example if a paddock has 0.3 grazing days, start_date + 0.3 == start_date. And all plots are based on start_date = end_date_of_previous_paddock, so errors accumulate down the table.

This is most easily seen in the non-growing season plan with 2 rotations through the paddocks in the same order. If you compare the end date of the recovery period, with its start of the second rotation the should match.

**Comment by mstenta** [comment]
Maybe the grazing days should just always be whole numbers? In practice, fractional days might not be necessary. I think the only time they would be is if you're moving the same herd more than once a day - which I don't think would ever happen in practice.

**Comment by woodbri** [comment]
Right be it is a matter of how we handle the math. For example what do we do with grazing a paddock for 0.3 grazing days. Regardless, we need to treat the problem consistently through out the code and the presentation and it is not a simple fix because all the code and math need to be review and updated.

**Comment by mstenta** [comment]
Yea that makes sense. Might be worth mapping out where all those calculations are, too, and trying to centralize/functionalize them somewhere to make future changes easier. Maybe that's already done to some extent - I need to review in more detail. You would know better than me.

**Comment by mstenta** [comment]
I'm running into this issue in my testing. The issue I'm seeing is that rotations are being generated with a duration of 0.

This is happening because the actual calculated grazing days are less than 0.5, and when it gets saved to the database (which only accepts whole numbers for the `duration` column), it is rounded down to zero.

I'm looking at the code that calculates the grazing days now, in `function farm_grazing_get_paddock_grazing_days()`. We could just add a little logic to the end of that to convert it to a non-zero whole number. Would that make sense?

**Comment by mstenta** [comment]
Something like this:

```
  // Round the days to a whole number.
  $days = round($days);

  // If it was rounded down to zero, enforce a minimum of 1 day.
  if ($days == 0) {
    $days = 1;
  }
```

That seems to work. Grazing days are always a whole number, and are never zero.

I'm not sure if that will adversely affect any other logic elsewhere in the module, though. But it seems to be working for the "Add rotation" and "Reset grazing days" logic.

**Comment by woodbri** [comment]
Yes, that makes sense to me, BUT if the herd is large enough or the paddock is small enough then this might not be a good thing, There is an expectation that the herd is on the paddock to graze, and if there is not enough forage there to feed the herd for a day this becomes a problem if it happens often during a plan. So maybe we need to generate an alert noting the fact when the grazing days are less then 0.75 for a paddock.

We might want to ask Paicines how they would deal with that.

**Comment by mstenta** [comment]
Yea that makes sense - agreed let's ask Paicines what they think.

**Comment by mstenta** [comment]
For now I added the logic to round (and ensure it's never zero) to the "rotations_form" branch: https://github.com/mstenta/farm_grazing/compare/rotations_form

**Comment by mstenta** [comment]
I need to amend my comment above (https://github.com/mstenta/farm_grazing/issues/4#issuecomment-361008588).

The reason the rounding was happening on my testing site was because the duration column in my database was an integer column - but the module's schema defines it as a float. So it was a bug on my system (not sure how that happened).

But in either case, it spurred good conversations anyway: https://github.com/mstenta/farm_grazing/issues/27
